libsupport/include/katana/Cache.h

```diff
@@ -0,0 +1,164 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_CACHE_H_
+#define KATANA_LIBSUPPORT_KATANA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+template <typename Key, typename Value>
+class KATANA_EXPORT Cache {
+  using ListType = std::list<Key>;
+  struct MapValue {
+    Value value;
+    // This allows us to delete the old position in the LRU list without a scan
+    typename ListType::iterator lru_it;
+  };
+  using MapType = std::unordered_map<Key, MapValue>;
+  enum class ReplacementPolicy { kLRUSize, kLRUBytes };
+
+public:
+  Cache(
+      size_t capacity,  // number of entries
+      std::function<void(const Key& key)> evict_cb = nullptr)
+      : policy_(ReplacementPolicy::kLRUSize),
+        capacity_(capacity),
+        value_to_bytes_(nullptr),
+        evict_cb_(std::move(evict_cb)) {
+    KATANA_LOG_VASSERT(capacity_ > 0, "cache requires positive capacity");
+  }
+  Cache(
+      size_t capacity,  // bytes of entries
+      std::function<void(const Key& key)> evict_cb,
+      std::function<size_t(const Value& value)> value_to_bytes)
+      : policy_(ReplacementPolicy::kLRUBytes),
+        capacity_(capacity),
+        value_to_bytes_(std::move(value_to_bytes)),
+        evict_cb_(std::move(evict_cb)) {
+    KATANA_LOG_VASSERT(capacity_ > 0, "cache requires positive capacity");
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUBytes || value_to_bytes_ != nullptr,
+        "kLRUBytes policy requires value to bytes function");
+  }
+
+  size_t size() const {
```

Comment:
`const`?

Reply:
Sorry, I don't get it.  It is a `const `method.